stores/slabbuffer.go

```diff
@@ -0,0 +1,505 @@
+package stores
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	rhpv2 "go.sia.tech/core/rhp/v2"
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/alerts"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/object"
+	"gorm.io/gorm"
+	"lukechampine.com/frand"
+)
+
+type SlabBuffer struct {
+	dbID     uint
+	filename string
+	slabKey  object.EncryptionKey
+	maxSize  int64
+
+	dbMu sync.Mutex
+
+	mu          sync.Mutex
+	file        *os.File
+	lockedUntil time.Time
+	size        int64
+	dbSize      int64
+	syncErr     error
+}
+
+type bufferGroupID [6]byte
+
+type SlabBufferManager struct {
+	bufferedSlabCompletionThreshold int64
+	dir                             string
+	s                               *SQLStore
+
+	mu                sync.Mutex
+	completeBuffers   map[bufferGroupID][]*SlabBuffer
+	incompleteBuffers map[bufferGroupID][]*SlabBuffer
+	buffersByKey      map[string]*SlabBuffer
+}
+
+func newSlabBufferManager(sqlStore *SQLStore, slabBufferCompletionThreshold int64, partialSlabDir string) (*SlabBufferManager, error) {
+	if slabBufferCompletionThreshold < 0 || slabBufferCompletionThreshold > 1<<22 {
+		return nil, fmt.Errorf("invalid slabBufferCompletionThreshold %v", slabBufferCompletionThreshold)
+	}
+
+	// load existing buffers
+	var buffers []dbBufferedSlab
+	err := sqlStore.db.
+		Joins("DBSlab").
+		Find(&buffers).
+		Error
+	if err != nil {
+		return nil, err
+	}
+	mgr := &SlabBufferManager{
+		bufferedSlabCompletionThreshold: slabBufferCompletionThreshold,
+		dir:                             partialSlabDir,
+		s:                               sqlStore,
+		completeBuffers:                 make(map[bufferGroupID][]*SlabBuffer),
+		incompleteBuffers:               make(map[bufferGroupID][]*SlabBuffer),
+		buffersByKey:                    make(map[string]*SlabBuffer),
+	}
+	for _, buffer := range buffers {
+		if buffer.DBSlab.ID == 0 {
+			// Buffer doesn't have a slab. We can delete it.
```

Comment:
when does this happen?  if it's unexpected should we add a log mention?

Reply:
Yeah it's pretty unexpected. I'll add a log line